write-fonts/src/font_builder.rs

```diff
@@ -15,6 +15,27 @@ pub struct FontBuilder<'a> {
     tables: BTreeMap<Tag, Cow<'a, [u8]>>,
 }
 
+impl TableDirectory {
+    pub fn from_table_records(table_records: Vec<TableRecord>) -> TableDirectory {
+        assert!(table_records.len() <= u16::MAX as usize);
+
+        // See https://learn.microsoft.com/en-us/typography/opentype/spec/otff#table-directory
+        // Computation works at the largest allowable num tables so don't stress the as u16's
+        let entry_selector = (table_records.len() as f64).log2().floor() as u16;
+        let search_range = (2.0_f64.powi(entry_selector as i32) * 16.0) as u16;
+        // The result doesn't really make sense with 0 tables but ... let's at least not fail
+        let range_shift = (table_records.len() * 16).saturating_sub(search_range as usize) as u16;
```

Comment:
i wonder, since we are doing this at least twice, here and in cmap table, maybe we should have a common method to compute these three fields?

Reply:
the other one is here:

https://github.com/googlefonts/fontations/blob/47cf723d23d1de419800e33c65eca918a2d75ccc/write-fonts/src/tables/cmap.rs#L29-L39

Reply:
they are basically the same, except for the item size, 16 in the case of the table directory, 2 in the case of cmap. FontTools uses the same `getSearchRange` function for both, parametrizing the itemSize.
https://github.com/fonttools/fonttools/blob/3b9a73ff8379ab49d3ce35aaaaf04b3a7d9d1655/Lib/fontTools/ttLib/ttFont.py#L1138-L1146